AzureAPIManagement/Services/ApiManagementClientBase.cs

```diff
@@ -0,0 +1,129 @@
+﻿using Newtonsoft.Json;
+using SmallStepsLabs.Azure.ApiManagement.Utilities;
+using System;
+using System.Diagnostics;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace SmallStepsLabs.Azure.ApiManagement
+{
+    public abstract class ApiManagementClientBase
+    {
+        private string Host { get; set; }
+
+
+        private string GetToken()
+        {
+            // Programmatically generate the access token used to call the API Management REST API.
+            // See "To programmatically create an access token" for instructions on how to get
+            // the values for id and key:
+            // https://msdn.microsoft.com/library/azure/5b13010a-d202-4af5-aabf-7ebc26800b3d#ProgrammaticallyCreateToken
+            // id - the value from the identifier text box in the credentials section of the
+            //      API Management REST API tab of the Security section.
+            string id = "<your identifier value here>";
+
+            // key - either the primary or secondary key from that same tab.
+            string key = "<either the primary or secondary key here>";
+            // expiry - the expiration date and time of the generated access token. In this example
+            //          the expiry is one day from the time the sample is run.
+
+            DateTime expiry = DateTime.UtcNow.AddDays(1);
+
+            // To programmatically create the access token so that we can authenticate and call the REST APIs,
+            // call this method. If you pasted in the access token from the publisher portal then you can
+            // comment out this line.
+            return Utility.CreateSharedAccessToken(id, key, expiry);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        protected HttpRequestMessage BuildRequest(object data = null)
+        {
+            StackTrace st = new StackTrace();
```

Comment:
This concerns me. I'd like to avoid Reflection code. Please implement this another way. If instead of an attribute you just have to hardcode this as a string in the method to pass as a parameter, that's fine. It's inline in the attribute anyways, no worse moving it a couple lines down.